instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,738 @@
+            )
+            setattr(
+                estimator, method_name, self.spanner(class_attr, estimator),
+            )
+
+    def _unwrap_function(self, function):
+        """Fetch the function underlying any decorators"""
+        if hasattr(function, "__wrapped__"):
+            return self._unwrap_function(function.__wrapped__)
+        return function
+
+    def _instrument_instance_method(
+        self,
+        estimator: BaseEstimator,
+        method_name: str,
+        attributes: Attributes = None,
+    ):
+        """Instrument an estimator instance method with a span.
+
+        When instrumenting we attach a tuple of (Class, method) to the
+        attribute ``_original_<method_name>`` for each method. This allows
+        us to replace the patched with the original in unstrumentation.
+
+        Args:
+            estimator (BaseEstimator): A fitted ``sklearn`` estimator.
+            method_name (str): The method name of the estimator on which to
+              apply a span.
+            attributes (dict): Attributes to attach to the spans.
+        """
+        if self._check_instrumented(estimator, method_name):
+            logger.debug(
+                "Already instrumented: %s.%s",
+                estimator.__class__.__qualname__,
+                method_name,
+            )
+            return
+
+        class_attr = getattr(type(estimator), method_name, None)
+        if isinstance(class_attr, property):
+            logger.debug(
+                "Not instrumenting found property: %s.%s",
+                estimator.__class__.__qualname__,
+                method_name,
+            )
+        else:
+            method = getattr(estimator, method_name)
+            setattr(
+                estimator, "_otel_original_" + method_name, (estimator, method)
+            )
+            setattr(
+                estimator,
+                method_name,
+                self.spanner(method, estimator, attributes),
+            )
+
+    def _instrument_estimator_attribute(
+        self, estimator: BaseEstimator, attributes: Attributes = None
+    ):
+        """Instrument instance attributes which also contain estimators.
+
+        Handle instance attributes which are also estimators, are a list
+        (Sequence) of estimators, or are mappings (dictionary) in which
+        the values are estimators.
+
+        Examples include ``RandomForestClassifier`` and
+        ``MultiOutputRegressor`` instances which have attributes
+        ``estimators_`` attributes.
+
+        Args:
+            estimator (BaseEstimator): A fitted ``sklearn`` estimator, with an
+              attribute which also contains an estimator or collection of
+              estimators.
+            attributes (dict): Attributes to attach to the spans.
+        """
+        attribs = self.recurse_attribs.get(estimator.__class__, [])
+        for attrib in attribs:
+            attrib_value = getattr(estimator, attrib)
```

Comment:
Existence check?

Reply:
I think we want to fail here with an `AttributeError` anyway, as it means the configuration is bad. Do we want to bubble up as something else?

Reply:
Hmm, that is true since this is happening during execution. Can we add a nicer error message? Would the `AttributeError` proide sufficient information for the user to troubleshoot do you think?

Reply:
I think it should be enough.